algebra/src/main/scala/endpoints/algebra/Requests.scala

```diff
@@ -0,0 +1,66 @@
+package endpoints.algebra
+
+import endpoints.Tupler
+
+import scala.language.higherKinds
+
+trait Requests extends Urls with Methods {
+
+  /** Information carried by requests’ headers */
+  type RequestHeaders[A]
+
+  /**
+    * No particular information. Does not mean that the headers *have to*
+    * be empty. Just that, from a server point of view no information will
+    * be extracted from them, and from a client point of view no particular
+    * headers will be built in the request.
+    */
+  def emptyHeaders: RequestHeaders[Unit]
+
+
+  /** Information carried by a whole request (headers and entity) */
+  type Request[A]
+
+  /** Information carried by request entity */
+  type RequestEntity[A]
+
+  /**
+    * Empty request.
+    */
+  def emptyRequest: RequestEntity[Unit]
+
+
+  /**
+    * Request for given parameters
+    *
+    * @param method Request method
+    * @param url Request URL
+    * @param entity Request entity
+    * @param headers Request headers
+    */
+  def request[A, B, C, AB](
+                            method: Method,
+                            url: Url[A],
+                            entity: RequestEntity[B] = emptyRequest,
+                            headers: RequestHeaders[C] = emptyHeaders
+                          )(implicit tuplerAB: Tupler.Aux[A, B, AB], tuplerABC: Tupler[AB, C]): Request[tuplerABC.Out]
```

Comment:
I’m a bit skeptical about this idea: I don’t have formal formatting rules (ie I have lots of exceptions).